### PR TITLE
Enable deployment tool to deploy new regions

### PR DESCRIPTION
**Pull request by mjudeikis**
Currently untested.  Should configure service key vault certs + secrets, ACR replication and DNS at RP deployment time.  Config format is as follows:

```yaml
rps:
- location: eastus
  subscriptionId: 00000000-0000-0000-0000-000000000000
  resourceGroupName: aro-eastus
  configuration:
    databaseAccountName: ''
    keyvaultPrefix: ''
    rpServerCertCommonName: ''
configuration:
  adminApiCaBundle: ''
  adminApiClientCertCommonName: ''
  clusterParentDomainName: ''
  extraCosmosDBIPs: ''
  extraKeyvaultAccessPolicies: []
  fpServicePrincipalId: ''
  globalMonitoringKeyVaultUri: ''
  globalResourceGroupName: ''
  globalSubscriptionId: ''
  mdmFrontendUrl: ''
  mdsdConfigVersion: ''
  mdsdEnvironment: ''
  pullSecret: ''
  rpImage: ''
  rpImageAuth: ''
  rpMode: ''
  rpParentDomainName: ''
  sshPublicKey: ''
```

### CODE REVIEW

File: pkg/deploy/predeploy.go [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (d *deployer) ensureServiceCertificates(ctx context.Context, serviceKeyVaultURI string) error {
+	_, err := d.keyvault.SetCertificateIssuer(ctx, serviceKeyVaultURI, "OneCert", keyvault.CertificateIssuerSetParameters{
```
Comment (mjudeikis) [inline pkg/deploy/predeploy.go]: Will this work during Pre-Deploy? I understood that this is only works after keyvault is being enable to work with OneCert?

File: pkg/deploy/predeploy.go [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+		},
+		{
+			certificateName: env.RPServerSecretName,
+			commonName:      d.config.Configuration.RPServerCertCommonName,
+			eku:             utilkeyvault.EkuServerAuth,
+		},
+	}
+
+	for _, c := range certs {
+		err = d.keyvault.CreateSignedCertificate(ctx, serviceKeyVaultURI, utilkeyvault.IssuerOnecert, c.certificateName, c.commonName, c.eku)
```
Comment (mjudeikis) [inline pkg/deploy/predeploy.go]:
```) keyvault.BaseClient#CreateCertificate: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="BadParameter" Message="Request to manage certificate is rejected with reason: Bad Request" 
exit status 1
```
When we enable oneCert, dont we need to wait for approval?

File: deploy/rp-development.json [hunk 3]
```diff
@@ -179,5 +179,15 @@
                 "[resourceId('Microsoft.Network/dnsZones', parameters('domainName'))]"
             ]
         }
-    ]
+    ],
+    "outputs": {
+        "[concat(parameters('domainName'), '-nameServers')]": {
```
Comment (mjudeikis) [inline deploy/rp-development.json]:
```
 The output name cannot be an expression. Please see https://aka.ms/arm-template/#outputs for usage details.'." AdditionalInfo=[{"info":{"lineNumber":1,"linePosition":211,"path":"properties.template.outputs['[concat(parameters(\\'domainName\\'), \\'-nameServers\\')]']"},"type":"TemplateViolation"}] 
```

File: pkg/deploy/predeploy.go (hunk 2 above)
Comment (JackQuincy) [reply pkg/deploy/predeploy.go]: No the approval for OneCert is per the user/principal requesting the cert be generated not per keyvault. You can put the issuer on any keyvault but only an pre-authorized user can generate a cert.

File: pkg/deploy/predeploy.go (hunk 1 above)
Comment (JackQuincy) [reply pkg/deploy/predeploy.go]: see other comment about OneCert cert generation works

**Comment by jim-minter** [comment]
@mj I'm removing your commit "add check for Sub Scope deployment".  I think you saw the error `RoleDefinitionWithSameNameExists: A role definition cannot be updated with a name that already exists.`, and I believe that was because you already had a RD with the name "ARO v4 ContainerRegistry Token Contributor" registered but with a different UUID.

I think it comes down to two RDs with two UUIDs can't share the same name.

I removed that RD from our subscription, and am now able to repeatedly PUT changes to the RD via an ARM template.

**Pull request merged**